comp/core/flare/flare.go

```diff
@@ -45,40 +45,41 @@ type dependencies struct {
 	Config        config.Component
 	IPCClient     ipcclient.Component `optional:"true"` // can be omitted in 'agent run'
 	Log           log.Component
-	Registrations []*Registration `group:"true"`
+	Registrations []registration `group:"true"`
 }
 
-type provides struct {
-	fx.Out
-
-	Component
-	IPCRoute *ipcserver.Route `group:"true"`
-}
-
-func newFlare(deps dependencies) provides {
+func newFlare(deps dependencies) (Component, ipcserver.Route) {
 	f := &flare{
-		registrations: deps.Registrations,
+		registrations: providedRegistrations(deps.Registrations),
 		ipcclient:     deps.IPCClient,
 		log:           deps.Log,
 	}
 
-	return provides{
-		Component: f,
-		IPCRoute:  ipcserver.NewRoute("/agent/flare", f.ipcHandler),
-	}
+	return f, ipcserver.NewRoute("/agent/flare", f.ipcHandler)
 }
 
 type mockDependencies struct {
 	fx.In
 
-	Registrations []*Registration `group:"true"`
+	Registrations []registration `group:"true"`
 }
 
 func newMock(deps mockDependencies) Component {
 	// mock is just like the real thing, but doesn't use ipcserver or config.
 	return &flare{
-		registrations: deps.Registrations,
+		registrations: providedRegistrations(deps.Registrations),
+	}
+}
+
+// providedRegistrations skips regsitrations with a nil callback.
+func providedRegistrations(registrations []registration) []registration {
+	provided := make([]registration, 0, len(registrations))
+	for _, r := range registrations {
+		if r.callback != nil {
```

Comment:
When it can happen?

Reply:
This can happen when the component that registered the flare callback decided not to do so.  It basically means that providing components have an "opt-out".  It's used for subscriptions when a component decides not to start and thus won't be consuming messages from the subscription.

Reply:
https://github.com/DataDog/dd-agent-comp-experiments/pull/17/files#diff-be2f02b435323adb19555016d85a562773a97bbe06b75f7c6b9986b0085ce22eR70